### ISSUE TITLE
Object destructuring type doesn't narrow parameter B when parameter A is narrowed

### ISSUE THREAD

**Issue by icecream17** [microsoft/TypeScript#44169]
# Bug Report

### 🔎 Search Terms

destructuring
narrowing
is not assignable to type

### 🕗 Version & Regression Information

(todo)

- This is a crash
- This changed between versions ______ and _______
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because _______

### 💻 Code

This works
```ts
type Dastructure = 
   {b: false, c: null} |
   {b: true, c: string}

class Example {
   prop?: string;
   constructor (options: Dastructure) {
      if (options.b) {
         this.prop = options.c
      }
   }
}
```

This gets an error
```ts
type Dastructure = 
   {b: false, c: null} |
   {b: true, c: string}

class Example {
   prop?: string;
   constructor ({b, c}: Dastructure) {
      if (b) {
         this.prop = c
//       ^^^^^^^^^
// Type 'string | null' is not assignable to type 'string | undefined'.
//   Type 'null' is not assignable to type 'string | undefined'.
      }
   }
}
```

### 🙁 Actual behavior
error

### 🙂 Expected behavior
not error?

I think it's because in the second code the context of ```{b, c}: A``` isn't considered when ```b``` is narrowed

**Comment by RyanCavanaugh** [comment]
This is a duplicate but I want to reraise the idea of CFA-linking variables sourced from the same destructuring expression

**Comment by dhythm** [comment]
I agree with this suggestion.
Type inference doesn't work well when I use destructuring.
I'm happy if this issue is solved. 🙏

![image](https://user-images.githubusercontent.com/10891318/119085745-42b81b80-ba3f-11eb-8197-d09ee5420597.png)

**Comment by remcovaes** [comment]
Is there any update on this? I am running into the exact same issue...

**Comment by icecream17** [comment]
This seems to be fixed by #46266
https://www.typescriptlang.org/play?ts=4.6.0-dev.20211117#code/C4TwDgpgBAIghgZ2AJwK4GNiudAvFAWACgpSBvAIwC4oAzOAGwQgBop0aA7VBhgXygAfYqSiUaKVK3Y0kyAJacA5n2LF0DRAigBRAB5wAtmAbQyI0mGQB7MAH5ZKRUoDcF9tc5yMwa8igAFJRs6Hw08N6Y2BAAlGLuovK0gRRx5iSimVDAABbyCAB0VrZQ+OgJpKoZUFV8QA
